### PR TITLE
Graph torch14 refactor

### PR DESCRIPTION
**Pull request by chicm-ms**
1.  Refactor model speedup graph part to use `torch.utils.tensorboard._pytorch_graph` data structures
2.  model speedup graph support 1.3.1 and 1.4.0
3.  Put nas graph and model speedup graph together 
4.  Model speedup unit test
5.  Upgrade pipeline pytorch and tensorflow version for model speedup unit test
6.  Support torchvision vgg16 (torch.flatten and AdaptiveAvgPool2d)

TODO:
1. special handle prim::GetAttr nodes for module inputs/outputs
2. retrieve node group inputs/outputs uniformly for module and nn.functional

### CODE REVIEW

File: src/sdk/pynni/tests/test_model_speedup.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+import math
+from unittest import TestCase, main
+
+from nni.graph_utils import build_module_graph
+from nni.compression.torch import L1FilterPruner
+from nni.compression.speedup.torch import ModelSpeedup
+
+class BackboneModel1(nn.Module):
```
Comment (ultmaster) [inline src/sdk/pynni/tests/test_model_speedup.py:16]: I think we should add tests for graph utils, as PyTorch is known to have issues on its tensorboard graph. For example, https://github.com/pytorch/pytorch/issues/33691 and https://github.com/pytorch/pytorch/issues/33670

File: src/sdk/pynni/nni/graph_utils.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+
+CLASSTYPE_KIND = 'ClassType'
+GETATTR_KIND = 'prim::GetAttr'
+
+_logger = logging.getLogger(__name__)
+
+def build_module_graph(model, dummy_input):
+    return TorchModuleGraph(model, dummy_input)
+
+def build_graph(model, dummy_input, verbose=False):
```
Comment (ultmaster) [inline src/sdk/pynni/nni/graph_utils.py:24]: graph_utils will be NNI's public API now?

File: src/sdk/pynni/tests/test_model_speedup.py (hunk 1 above)
Comment (ultmaster) [reply src/sdk/pynni/tests/test_model_speedup.py:16]: Maybe we can "borrow" their tests: https://github.com/pytorch/pytorch/blob/253943d5a7bb90a420e4d94366101915823c7929/test/test_tensorboard.py#L505

Comment (chicm-ms) [reply src/sdk/pynni/tests/test_model_speedup.py:15]: Good suggestion,  added the borrowed test case from pytorch v1.4.

File: src/sdk/pynni/nni/graph_utils.py (hunk 2 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/graph_utils.py:24]: It is just an utility, I would not say it a public API.

File: src/sdk/pynni/tests/test_graph_utils.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    def __init__(self):
+        super().__init__()
+        self.backbone1 = BackboneModel1()
+        self.backbone2 = BackboneModel2()
+        self.fc3 = nn.Linear(10, 2) 
+    def forward(self, x):
+        x = self.backbone1(x)
+        x = self.backbone2(x)
+        x = self.fc3(x)
+        return x
```
Comment (ultmaster) [inline src/sdk/pynni/tests/test_graph_utils.py:57]: I'd like two more models to be added as the graph utils provided by PyTorch fail for them.

```python
class MyModule(nn.Module):
  def __init__(self):
    super().__init__()
    self.weight = nn.Linear(5, 3)
    self.bias = nn.Linear(5, 3)
    self.module = nn.Linear(6, 1)

  def forward(self, x):
    tensors = [self.weight(x), self.bias(x)]
    self.module(torch.cat(tensors, dim=1))
    return x
```

```python
class MyModule(nn.Module):
  def __init__(self):
    super().__init__()
    self.module = nn.ModuleList([
      nn.Linear(5, 3),
      nn.Linear(3, 1)
    ])

  def forward(self, x):
    x = self.module[0](x)
    x = self.module[1](x)
    return x
```

File: src/sdk/pynni/nni/graph_utils.py (hunk 2 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/graph_utils.py:24]: In that case, better rename it as `_graph_utils.py`

File: src/sdk/pynni/tests/test_graph_utils.py (hunk 3 above)
Comment (chicm-ms) [reply src/sdk/pynni/tests/test_graph_utils.py:57]: ok, can I save the output of your code as expected file?

File: src/sdk/pynni/nni/graph_utils.py (hunk 2 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/graph_utils.py:24]: ok

File: src/sdk/pynni/tests/test_graph_utils.py (hunk 3 above)
Comment (ultmaster) [reply src/sdk/pynni/tests/test_graph_utils.py:57]: Sure.

First one:

Code:

```python
class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = nn.Linear(5, 3)
        self.bias = nn.Linear(5, 3)
        self.module = nn.Linear(6, 1)

    def forward(self, x):
        tensors = [self.weight(x), self.bias(x)]
        self.module(torch.cat(tensors, dim=1))
        return x
```

Expected result:

```
node {
  name: "input/input.1"
  op: "IO Node"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 5
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: ""
    }
  }
}
node {
  name: "output/output.1"
  op: "IO Node"
  input: "input/input.1"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 5
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: ""
    }
  }
}
node {
  name: "MyModule/Linear[weight]/bias/49"
  op: "prim::GetAttr"
  input: "MyModule/Linear[weight]/weight/35"
  attr {
    key: "attr"
    value {
      s: "{ name :  bias }"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/weight/50"
  op: "prim::GetAttr"
  input: "MyModule/Linear[weight]/weight/35"
  attr {
    key: "attr"
    value {
      s: "{ name :  weight }"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/51"
  op: "aten::t"
  input: "MyModule/Linear[weight]/weight/50"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 5
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/52"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/53"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/54"
  op: "aten::addmm"
  input: "MyModule/Linear[weight]/bias/49"
  input: "input/input.1"
  input: "MyModule/Linear[weight]/51"
  input: "MyModule/Linear[weight]/52"
  input: "MyModule/Linear[weight]/53"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/bias/55"
  op: "prim::GetAttr"
  input: "MyModule/Linear[bias]/weight/38"
  attr {
    key: "attr"
    value {
      s: "{ name :  bias }"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/weight/56"
  op: "prim::GetAttr"
  input: "MyModule/Linear[bias]/weight/38"
  attr {
    key: "attr"
    value {
      s: "{ name :  weight }"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/57"
  op: "aten::t"
  input: "MyModule/Linear[bias]/weight/56"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 5
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/58"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/59"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/60"
  op: "aten::addmm"
  input: "MyModule/Linear[bias]/bias/55"
  input: "input/input.1"
  input: "MyModule/Linear[bias]/57"
  input: "MyModule/Linear[bias]/58"
  input: "MyModule/Linear[bias]/59"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/23"
  op: "prim::ListConstruct"
  input: "MyModule/Linear[weight]/54"
  input: "MyModule/Linear[bias]/60"
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/24"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/input"
  op: "aten::cat"
  input: "MyModule/23"
  input: "MyModule/24"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 6
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/61"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
versions {
  producer: 22
}
```

Comment (ultmaster) [reply src/sdk/pynni/tests/test_graph_utils.py:57]: Second one:

Code:

```python
class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.module = nn.ModuleList([
            nn.Linear(5, 3),
            nn.Linear(3, 1)
        ])

    def forward(self, x):
        x = self.module[0](x)
        x = self.module[1](x)
        return x
```

Expected result:

```
node {
  name: "input/input.1"
  op: "IO Node"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 5
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: ""
    }
  }
}
node {
  name: "output/output.1"
  op: "IO Node"
  input: "MyModule/ModuleList[module]/Linear[1]/46"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 1
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: ""
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/bias/35"
  op: "prim::GetAttr"
  input: "MyModule/ModuleList[module]/Linear[0]/weight/26"
  attr {
    key: "attr"
    value {
      s: "{ name :  bias }"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/weight/36"
  op: "prim::GetAttr"
  input: "MyModule/ModuleList[module]/Linear[0]/weight/26"
  attr {
    key: "attr"
    value {
      s: "{ name :  weight }"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/37"
  op: "aten::t"
  input: "MyModule/ModuleList[module]/Linear[0]/weight/36"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 5
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/38"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/39"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/input"
  op: "aten::addmm"
  input: "MyModule/ModuleList[module]/Linear[0]/bias/35"
  input: "input/input.1"
  input: "MyModule/ModuleList[module]/Linear[0]/37"
  input: "MyModule/ModuleList[module]/Linear[0]/38"
  input: "MyModule/ModuleList[module]/Linear[0]/39"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/bias/41"
  op: "prim::GetAttr"
  input: "MyModule/ModuleList[module]/Linear[1]/weight/30"
  attr {
    key: "attr"
    value {
      s: "{ name :  bias }"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/weight/42"
  op: "prim::GetAttr"
  input: "MyModule/ModuleList[module]/Linear[1]/weight/30"
  attr {
    key: "attr"
    value {
      s: "{ name :  weight }"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/43"
  op: "aten::t"
  input: "MyModule/ModuleList[module]/Linear[1]/weight/42"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 3
          }
          dim {
            size: 1
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/44"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/45"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/46"
  op: "aten::addmm"
  input: "MyModule/ModuleList[module]/Linear[1]/bias/41"
  input: "MyModule/ModuleList[module]/Linear[0]/input"
  input: "MyModule/ModuleList[module]/Linear[1]/43"
  input: "MyModule/ModuleList[module]/Linear[1]/44"
  input: "MyModule/ModuleList[module]/Linear[1]/45"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 1
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
versions {
  producer: 22
}
```

Comment (ultmaster) [reply src/sdk/pynni/tests/test_graph_utils.py:57]: Test them with:

```python
model = MyModule()
graph_def, _ = graph(model, torch.randn(4, 5), verbose=False)
result = text_format.MessageToString(graph_def)

print(result)
```

File: azure-pipelines.yml [hunk 4]
```diff
@@ -26,9 +26,9 @@ jobs:
       yarn eslint
     displayName: 'Run eslint'
   - script: |
-      python3 -m pip install torch==1.2.0 --user
-      python3 -m pip install torchvision==0.4.0 --user
-      python3 -m pip install tensorflow==1.13.1 --user
+      python3 -m pip install torch==1.3.1 --user
```
Comment (ultmaster) [inline azure-pipelines.yml:29]: Can we know for sure whether it works on torch 1.4 and torch 1.5?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+import logging
+import queue
+import re
+from collections import defaultdict
+import torch
+from torch.utils.tensorboard._pytorch_graph import NodePy, NodePyIO, NodePyOP, GraphPy
+from tensorboard.compat.proto.config_pb2 import RunMetadata
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from tensorboard.compat.proto.step_stats_pb2 import StepStats, DeviceStepStats
+from tensorboard.compat.proto.versions_pb2 import VersionDef
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:14]: which package is this `tensorboard` from? tensorflow?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+        self.bound_model = model
+        self._trace(model, dummy_input)
+
+    def _trace(self, model, dummy_input):
+        with torch.onnx.set_training(model, False):
+            self.trace = torch.jit.trace(model, dummy_input)
+            torch._C._jit_pass_inline(self.trace.graph)
+
+class TorchProtoGraph(TorchGraph):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:58]: better to add docstring for this class

File: src/sdk/pynni/nni/_graph_utils.py [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+        Args:
+        graph (PyTorch module): The model graph to be parsed.
+        trace (PyTorch JIT TracedModule): The model trace to be parsed.
+        args (tuple): input tensor[s] for the model.
+        omit_useless_nodes (boolean): Whether to remove nodes from the graph.
+        """
+        nodes_py = GraphPy()
+        for node in graph.inputs():
+            if omit_useless_nodes:
+                if not node.uses():  # number of user of the node (= number of outputs/ fanout)
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:81]: number of `user`?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 8]
```diff
@@ -0,0 +1,446 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:1]: It seems you rewrite the logic of `graph_utils.py` in pytorch. just wondering is it possible to directly use the output of `graph_utils.py` in pytorch?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+            module_name = ''
+            for i, alias in enumerate(module_aliases):
+                if i == 0:
+                    module_name = alias
+                    node.scopeName = base_name
+                else:
+                    module_name += '.' + alias
+                    node.scopeName += '/' + (alias_to_name[module_name] if module_name in alias_to_name else alias)
+
+        nodes_py.populate_namespace_from_OP_to_IO()
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:132]: what does this API do?

File: src/sdk/pynni/nni/_graph_utils.py [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+    def sub_node_names(self):
+        return [x.name for x in self.nodes]
+
+    def __repr__(self):
+        return 'name: {}, type: {}, op_type: {}, sub_nodes: {}, inputs: {}, outputs: {}'.format(
+            self.name, self.type, self.op_type, self.sub_node_names(), self.inputs, self.outputs
+        )
+
+
+class TorchModuleGraph(TorchGraph):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:163]: better to add docstring for this class

File: src/sdk/pynni/nni/_graph_utils.py [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+                    module_name = alias
+                    node.scopeName = base_name
+                else:
+                    module_name += '.' + alias
+                    node.scopeName += '/' + (alias_to_name[module_name] if module_name in alias_to_name else alias)
+
+        nodes_py.populate_namespace_from_OP_to_IO()
+        return nodes_py.to_proto()
+
+class NodePyGroup(NodePy):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:135]: better to add docstring for this class and its member functions

File: src/sdk/pynni/nni/_graph_utils.py (hunk 8 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/_graph_utils.py:1]: each node in graph_utils.py is directly mapped a single traced C node, the main logic of the graph for shape inference is focusing on grouping multiple nodes together.  Directly use the output of graph_utils.py introduces additional dependencies and save no efforts.

File: src/sdk/pynni/nni/_graph_utils.py (hunk 7 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/_graph_utils.py:90]: This is from pytorch implementation,  means number of references.

File: src/sdk/pynni/nni/_graph_utils.py (hunk 6 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/_graph_utils.py:54]: Thanks, fixed.

File: src/sdk/pynni/nni/_graph_utils.py (hunk 10 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/_graph_utils.py:193]: Thanks, docstring added.

File: src/sdk/pynni/nni/_graph_utils.py (hunk 11 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/_graph_utils.py:144]: Thanks, docstring added.

File: src/sdk/pynni/nni/_graph_utils.py (hunk 5 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/_graph_utils.py]: Yes, from tensorflow, moved.

File: azure-pipelines.yml (hunk 4 above)
Comment (chicm-ms) [reply azure-pipelines.yml]: Added a separate pipeline for latest torch and v1.3.1

File: src/sdk/pynni/nni/_graph_utils.py (hunk 9 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/_graph_utils.py:141]: It builds inputs/outputs connection between nodes by creating io nodes from op nodes.
source code can be found here: 
https://github.com/pytorch/pytorch/blob/master/torch/utils/tensorboard/_pytorch_graph.py#L142

**Comment by QuanluZhang** [comment]
please update doc to remove the limitation of only supporting pytorch 1.3.1

File: src/sdk/pynni/nni/_graph_utils.py [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+            if len(parts1) >= len(parts2):
+                return False
+            for i in range(len(parts1)):
+                if parts2[i] != parts1[i]:
+                    return False
+            return True
+        module_names = sorted([x[0] for x in self.trace.named_modules() if x[0]])
+        leaf_nodes = []
+        for i, name in enumerate(module_names):
+            if i + 1 >= len(module_names) or not is_parent(name, module_names[i + 1]):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/_graph_utils.py:328]: what is the meaning of the first condition?

Comment (QuanluZhang) [reply src/sdk/pynni/nni/_graph_utils.py:328]: got it.

**Pull request merged**